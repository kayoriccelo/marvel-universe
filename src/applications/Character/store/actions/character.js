import api from '../../../../services/api'
import * as Types from '../types'
import history from '../../../../history'


export const getListingCharactersAPI = params => {
    let search = params['search'] ? `?nameStartsWith=${params['search']}` : ''

    return dispatch => {
        return api.get(`/v1/public/characters${search}`).then(response => {
            dispatch({ type: Types.LISTING_CHARACTERS, payload: response.data.data.results })

            return response.data.results
        })
    }
}

export const getListingCharacters = _ => {
    return dispatch => dispatch({ type: Types.LISTING_CHARACTERS })
}

export const getListingSeries = character => {
    return dispatch => {
        return api.get(`/v1/public/characters/${character}/series`).then(response => {
            let series = response.data.data ? response.data.data.results : response.data

            dispatch({ type: Types.LISTING_SERIES, payload: series })

            return series
        })
    }
}

export const clearSeries = _ => {
    return dispatch => {
        dispatch({ type: Types.LISTING_SERIES, payload: [] })
    }
}

export const loadCharacter = params => {
    return dispatch => {
        return api.get(`/v1/public/characters/${params['id']}`).then(response => {
            if (response.data.data !== undefined) {
                let character = response.data.data.results[0]
                dispatch([{ type: Types.LOAD_CHARACTER, payload: character }, getListingSeries(character.id)])
            } else {
                let character = response.data
                dispatch({ type: Types.LOAD_CHARACTER, payload: character })
            }
        })
    }
}

export const updateCharacter = params => {
    return dispatch => {
        let itens = params['itens'].map(item => item.id === params['instance'].id ? params['instance'] : item)

        dispatch([
            { type: Types.UPDATE_CHARACTER, payload: params['instance'] },
            { type: Types.LISTING_CHARACTERS, payload: itens }
        ])

        history.push('/characters/listing')
    }
}