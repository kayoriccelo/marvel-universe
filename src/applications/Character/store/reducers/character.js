import * as Types from '../types'


export const initialState = {
    itens: [],
    series: [],
    instance: null
}

export const characterReducer = (state = initialState, action) => {
    switch (action.type) {

        case Types.LISTING_CHARACTERS:
            return { ...state, itens: action.payload || state.itens }

        case Types.LOAD_CHARACTER:
            return { ...state, instance: action.payload }

        case Types.UPDATE_CHARACTER:
            return { ...state, instance: action.payload }

        case Types.LISTING_SERIES:
            return { ...state, series: action.payload }

        default:
            return state
    }
}
