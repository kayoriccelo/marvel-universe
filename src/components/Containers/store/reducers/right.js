import * as Types from '../types'


const initialState = {
    title: null
}

export const rightReducer = (state = initialState, action) => {
    switch (action.type) {

        case Types.SET_TITLE_BREADCRUMBS:
            return { ...state, title: action.payload }

        default:
            return state
    }
}
