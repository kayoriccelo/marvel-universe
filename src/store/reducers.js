import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import { rightReducer } from '../components/Containers/store/reducers'
import { characterReducer } from '../applications/Character/store/reducers'


const reducers = combineReducers({
    router: routerReducer,
    right: rightReducer,
    character: characterReducer
})

export default reducers
