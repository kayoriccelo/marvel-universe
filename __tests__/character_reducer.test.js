import { characterReducer, initialState } from '../src/applications/Character/store/reducers/character'

describe('Character reducer', () => {
    it('should return the initial state', () => {
        expect(characterReducer(undefined, {})).toMatchSnapshot()
    })

    it('should return the itens "LISTING_CHARACTERS"', () => {
        expect(characterReducer(initialState, { type: 'LISTING_CHARACTERS' })).toMatchSnapshot()
    })

    it('should return the instance "LOAD_CHARACTER"', () => {
        expect(characterReducer(initialState, { type: 'LOAD_CHARACTER' })).toMatchSnapshot()
    })

    it('should return the instance updated "UPDATE_CHARACTER"', () => {
        const { instance } = initialState
        expect(characterReducer(initialState, { type: 'UPDATE_CHARACTER', payload: {...instance, name: 'Testing'} })).toMatchSnapshot()
    })

    it('should return the series "LISTING_SERIES"', () => {
        expect(characterReducer(initialState, { type: 'LISTING_SERIES' })).toMatchSnapshot()
    })
})