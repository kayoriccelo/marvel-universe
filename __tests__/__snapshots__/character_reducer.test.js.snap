// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character reducer should return the initial state 1`] = `
Object {
  "instance": null,
  "itens": Array [],
  "series": Array [],
}
`;

exports[`Character reducer should return the instance "LOAD_CHARACTER" 1`] = `
Object {
  "instance": undefined,
  "itens": Array [],
  "series": Array [],
}
`;

exports[`Character reducer should return the instance updated "UPDATE_CHARACTER" 1`] = `
Object {
  "instance": Object {
    "name": "Testing",
  },
  "itens": Array [],
  "series": Array [],
}
`;

exports[`Character reducer should return the itens "LISTING_CHARACTERS" 1`] = `
Object {
  "instance": null,
  "itens": Array [],
  "series": Array [],
}
`;

exports[`Character reducer should return the series "LISTING_SERIES" 1`] = `
Object {
  "instance": null,
  "itens": Array [],
  "series": undefined,
}
`;
